connectionpool.go

```
@@ -435,3 +439,404 @@ func (c *SimplePool) removeHostLocked(addr string) {
+}
+
+func newHostConnPool(
+	host string,
+	port int,
+	size int,
+	connCfg ConnConfig,
+	keyspace string,
+	policy ConnSelectionPolicy,
+) *hostConnPool {
+
+	pool := &hostConnPool{
+		host:     host,
+		port:     port,
+		addr:     JoinHostPort(host, port),
+		size:     size,
+		connCfg:  connCfg,
+		keyspace: keyspace,
+		policy:   policy,
+		conns:    make([]*Conn, 0, size),
+		filling:  false,
+		closed:   false,
+	}
+
+	// fill the pool with the initial connections before returning
+	pool.fill()
+
+	return pool
+}
+
+// Pick a connection from this connection pool for the given query.
+func (pool *hostConnPool) Pick(qry *Query) *Conn {
+	pool.mu.Lock()
+	if pool.closed {
+		pool.mu.Unlock()
+		return nil
+	}
+
+	empty := len(pool.conns) == 0
+	pool.mu.Unlock()
+
+	if empty {
+		// try to fill the empty pool
+		pool.fill()
+	}
+
+	return pool.policy.Pick(qry)
+}
+
+//Size returns the number of connections currently active in the pool
+func (pool *hostConnPool) Size() int {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	return len(pool.conns)
+}
+
+//Close the connection pool
+func (pool *hostConnPool) Close() {
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+
+	if pool.closed {
+		return
+	}
+	pool.closed = true
+
+	// drain, but don't wait
+	go pool.drain()
+}
+
+// Fill the connection pool
+func (pool *hostConnPool) fill() {
+	pool.mu.Lock()
+	// avoid filling a closed pool, or concurrent filling
+	if pool.closed || pool.filling {
+		pool.mu.Unlock()
+		return
+	}
+
+	// determine the filling work to be done
+	startCount := len(pool.conns)
+	fillCount := pool.size - startCount
+
+	// avoid filling a full (or overfull) pool
+	if fillCount <= 0 {
+		pool.mu.Unlock()
+		return
+	}
+
+	// ok fill the pool
+	pool.filling = true
+
+	// allow others to access the pool while filling
+	pool.mu.Unlock()
+
+	// fill only the first connection synchronously
+	if startCount == 0 {
+		err := pool.connect()
+		if opErr, ok := err.(*net.OpError); ok && opErr.Op == "read" {
```

This check at this line is meant to avoid logging errors typically seen when the pool is trying to connect to a host which is down or unreachable since the pool attempts to connect so often (though less often than SimplePool). For net.Dial and net.DialTimeout mostly it seems like net.OpError is returned for most issues, such as "connection refused".

On reviewing this code while running the connectionpool_systems_test.sh I noticed that I was not suppressing errors with Op == "dial" so I'll be fixing that.